### PR TITLE
[Api] Add test app for ApiBundle

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

Tests for adding new api resource for new entities: 
- Not sylius resource entity: `Foo`
- Sylius resource entity: `FooSyliusResource`
- Extended sylius resource entity: `Taxon`

Tests for overwriting Sylius ApiPlatform's resource entities : `Promotion`

To test locally, run commands:
`(cd src/Sylius/Bundle/ApiBundle && composer install)`
`(cd src/Sylius/Bundle/ApiBundle/test && bin/console doctrine:database:create -e test)`
`(cd src/Sylius/Bundle/ApiBundle/test && bin/console doctrine:schema:update --force -e test)`
`(cd src/Sylius/Bundle/ApiBundle/test && bin/console assets:install public)`
`(cd src/Sylius/Bundle/ApiBundle/test && APP_ENV=test symfony serve)`

To run tests:
`(cd src/Sylius/Bundle/ApiBundle && vendor/bin/phpunit)`

Swagger is working properly:
Screenshot 2021-02-10 at 16 19 26
Screenshot 2021-02-11 at 09 08 00

### CODE REVIEW

File: .github/workflows/packages.yml [hunk 1]
```diff
@@ -100,7 +100,15 @@ jobs:
...
                 name: Test package
+#                if: ${{ matrix.package }} != ApiBundle
                 run: bin/test-package src/Sylius/${{ matrix.package }}
+                
+#            -
+#                name: Test ApiBundle
+#                if: ${{ matrix.package }} == ApiBundle
+#                run: |
+#                    (cd src/Sylius/Bundle/ApiBundle/test && bin/console doctrine:schema:update --force -e test)
+#                    bin/test-package src/Sylius/ApiBundle
```
Comment (GSadee) [inline .github/workflows/packages.yml:111]: I guess that this change should be reverted

File: src/Sylius/Bundle/ApiBundle/test/tests/FooSyliusResourceTest.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    public function setUp(): void
+    {
+        $this->setFixturesFiles(['test/fixtures/foo_sylius_resource.yaml']);
+        $this->setUpTest();
+    }
+
+    /**
+     * @test
+     */
+    public function it_allows_to_get_collection_as_a_login_administrator_on_new_not_admin_resource(): void
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/test/tests/FooSyliusResourceTest.php:31]:
```suggestion
    public function it_allows_to_get_collection_as_a_logged_in_administrator_on_new_not_admin_resource(): void
```

**Review by GSadee — COMMENTED** [review]
Shouldn't this PR be opened to 1.9 branch?

File: src/Sylius/Bundle/ApiBundle/test/config/packages/_sylius.yaml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        entity_managers:
+            default:
+                mappings:
+                    test:
+                        mapping:              true
+                        type:                 xml
+                        dir:                  "%kernel.project_dir%/src/Resources/config/doctrine"
+                        alias:                ~
+                        prefix:               Sylius\Bundle\ApiBundle\test\src\Entity
+                        is_bundle:            false
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/test/config/packages/_sylius.yaml:41]:
```suggestion
                        mapping: true
                        type: xml
                        dir: "%kernel.project_dir%/src/Resources/config/doctrine"
                        alias: ~
                        prefix: Sylius\Bundle\ApiBundle\test\src\Entity
                        is_bundle: false
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Is this PR fixes https://github.com/Sylius/Sylius/issues/12152 ?

**Comment by stloyd** [comment]
> Is this PR fixes #12152 ?

No, this is just confirming that using Sylius + Api-platform works, but issue is about overwriting/extending/injecting Sylius entities/resources, which is not covered here & requires mentioned changes in the issue (or even more).

**Comment by Tomanhez** [comment]
Hi @stloyd 

I tested overwriting resource configuration on `Promotion` entities and its look working properly. I will add test for extended sylius resource with overwrited api resource configuration.

File: src/Sylius/Bundle/ApiBundle/composer.json [hunk 4]
```diff
@@ -32,7 +32,17 @@
         "matthiasnoback/symfony-config-test": "^4.2",
         "matthiasnoback/symfony-dependency-injection-test": "^4.1",
         "phpspec/phpspec": "^7.0",
-        "phpunit/phpunit": "^8.5"
+        "phpunit/phpunit": "^8.5",
+        "symfony/monolog-bundle": "^3.5",
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/composer.json:36]: why do we need it?

File: src/Sylius/Bundle/ApiBundle/composer.json [hunk 5]
```diff
@@ -32,7 +32,17 @@
...
+        "symfony/debug-bundle": "^4.4 || ^5.2",
+        "friends-of-behat/symfony-extension": "^2.1",
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/composer.json:38]: Do we run any behat tests in this bundle?

File: src/Sylius/Bundle/ApiBundle/composer.json [hunk 6]
```diff
@@ -32,7 +32,17 @@
...
+        "lexik/jwt-authentication-bundle": "^2.6",
+        "theofidry/alice-data-fixtures": "^1.4",
+        "nelmio/alice": "^3.6",
+        "polishsymfonycommunity/symfony-mocker-container": "^1.0",
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/composer.json:42]: Have you mocked anything here?

File: src/Sylius/Bundle/ApiBundle/test/.env [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# This file is a "template" of which env vars needs to be defined in your configuration or in an .env file
+# Set variables here that may be different on each deployment target of the app, e.g. development, staging, production.
+# https://symfony.com/doc/current/best_practices/configuration.html#infrastructure-related-configuration
+
+###> symfony/framework-bundle ###
+APP_ENV=dev
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/test/.env:6]: I'm wondering, if we can't just skip these files and hardcode all these values. It doesn't have to be modified we will use this app only for test purposes

File: src/Sylius/Bundle/ApiBundle/test/composer.json [hunk 8]
```diff
@@ -0,0 +1,3 @@
+{
+    "name": "example/api-bundle-test-application"
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/test/composer.json:2]: Why do we need this file?

File: src/Sylius/Bundle/ApiBundle/test/config/packages/_sylius.yaml [hunk 9]
```diff
@@ -0,0 +1,94 @@
+imports:
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/test/config/packages/_sylius.yaml:1]: Should this file be called `_sylius`? I can see all kinds of configs here despite sylius one. Let's rename this file to `config.yaml`. Also, can you put config of `api_platform.yaml` and `security.yaml` here as well?

File: src/Sylius/Bundle/ApiBundle/test/src/AppKernel.php [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        $loader->load($confDir . '/{services}_' . $this->environment . self::CONFIG_EXTS, 'glob');
+    }
+
+    protected function configureRoutes(RouteCollectionBuilder $routes): void
+    {
+        $confDir = $this->getProjectDir() . '/config';
+
+        $routes->import($confDir . '/{routes}/*' . self::CONFIG_EXTS, '/', 'glob');
+        $routes->import($confDir . '/{routes}/' . $this->environment . '/**/*' . self::CONFIG_EXTS, '/', 'glob');
+        $routes->import($confDir . '/{routes}' . self::CONFIG_EXTS, '/', 'glob');
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/test/src/AppKernel.php:81]: Can we simplify this config to load only a particular file?

File: src/Sylius/Bundle/ApiBundle/test/src/AppKernel.php [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    protected function getContainerBaseClass(): string
+    {
+        if ($this->isTestEnvironment()) {
+            return MockerContainer::class;
+        }
+
+        return parent::getContainerBaseClass();
+    }
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/test/src/AppKernel.php:91]: Probably can be removed

File: src/Sylius/Bundle/ApiBundle/test/src/AppKernel.php [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+            new ClosureLoader($container),
+        ));
+
+        return new DelegatingLoader($resolver);
+    }
+
+    private function isTestEnvironment(): bool
+    {
+        return 0 === strpos($this->getEnvironment(), 'test');
+    }
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/test/src/AppKernel.php:115]: Can be removed

File: src/Sylius/Bundle/ApiBundle/test/fixtures/administrator.yaml [hunk 13]
```diff
@@ -0,0 +1,8 @@
+Sylius\Component\Core\Model\AdminUser:
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/test/fixtures/administrator.yaml:1]: I would say, that this file should be placed in the config folder -> `src/Sylius/Bundle/ApiBundle/test/config/fixtures/administrator.yaml`

File: src/Sylius/Bundle/ApiBundle/test/tests/FooSyliusResourceTest.php [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\test\tests;
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/test/tests/FooSyliusResourceTest.php:14]: This namespace shows, that something is odd with this config. Can we do something like this
`Sylius\Bundle\ApiBundle\test\` <- main folder for test app
`Sylius\Bundle\ApiBundle\test\Application` <- main folder for test app
`Sylius\Bundle\ApiBundle\test\Application\config` <- main folder for test app 
`Sylius\Bundle\ApiBundle\test\Application\src` <- main folder for test code
`Sylius\Bundle\ApiBundle\test\Tests` <- main folder for tests classes

File: src/Sylius/Bundle/ApiBundle/test/tests/Test.php [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\ApiBundle\test\tests;
+
+use Fidry\AliceDataFixtures\LoaderInterface;
+use Fidry\AliceDataFixtures\Persistence\PurgeMode;
+use Lexik\Bundle\JWTAuthenticationBundle\Services\JWTTokenManagerInterface;
+
+trait Test
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/test/tests/Test.php:18]: What kind of `test` is this test? Can you give a more meaningful name?

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Bridge/Symfony/Routing/RouteNameResolver.php [hunk 16]
```diff
@@ -89,12 +89,20 @@ private function returnMatchingRouteName(
         }
 
         foreach ($matchingRoutes as $routeName => $route) {
-            $requestPrefix = $this->pathPrefixProvider->getCurrentPrefix();
             $routePrefix = $this->pathPrefixProvider->getPathPrefix($route->getPath());
+            $requestPrefix = $this->pathPrefixProvider->getCurrentPrefix();
+
+            if ($requestPrefix === '' && str_contains($routeName, '_shop_')) {
```
Comment (stloyd) [inline src/Sylius/Bundle/ApiBundle/ApiPlatform/Bridge/Symfony/Routing/RouteNameResolver.php:95]: Are we sure to use PHP 8 functions?

Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/ApiPlatform/Bridge/Symfony/Routing/RouteNameResolver.php]: Nice catch 👍

File: src/Sylius/Bundle/ApiBundle/spec/Provider/PathPrefixProviderSpec.php [hunk 17]
```diff
@@ -79,6 +79,6 @@ function it_returns_shop_prefix_if_there_is_no_logged_in_user(UserContextInterfa
     {
         $userContext->getUser()->willReturn(null);
 
-        $this->getCurrentPrefix()->shouldReturn('shop');
+        $this->getCurrentPrefix()->shouldReturn('');
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/Provider/PathPrefixProviderSpec.php:82]: I assume, that the name of this test should be adjusted. Also, I'm wondering if it changes the current behavior. Have you checked it?

File: src/Sylius/Bundle/ApiBundle/test/config/packages/config.yaml [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+            classes:
+                model: Sylius\Bundle\ApiBundle\test\src\Entity\Taxon
+
+sylius_resource:
+    resources:
+        test.foo_sylius_resource:
+            classes:
+                model: Sylius\Bundle\ApiBundle\test\src\Entity\FooSyliusResource
+                
+
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/test/config/packages/config.yaml:27]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/ApiPlatform/Bridge/Symfony/Routing/RouteNameResolver.php [hunk 19]
```diff
@@ -89,12 +89,20 @@ private function returnMatchingRouteName(
         }
 
         foreach ($matchingRoutes as $routeName => $route) {
-            $requestPrefix = $this->pathPrefixProvider->getCurrentPrefix();
             $routePrefix = $this->pathPrefixProvider->getPathPrefix($route->getPath());
+            $requestPrefix = $this->pathPrefixProvider->getCurrentPrefix();
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/ApiPlatform/Bridge/Symfony/Routing/RouteNameResolver.php:93]: Can you remind me, how do we test this behavior?

File: src/Sylius/Bundle/ApiBundle/spec/Provider/PathPrefixProviderSpec.php (hunk 17 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/spec/Provider/PathPrefixProviderSpec.php:82]: All behats are green :) In our api we have three types users `AdminUser`, `ShopUser`, `Visitor` and two prefix types `admin` , `shop`. We assumed that Visitor will be using only `shop` prefix, but what if someone adds new endpoint without prefix? This changes must be added to properly work with changes on `RouteNameResolver.php`. This provider is used in all requests on tests and all test are fine. We have only this test for it.

File: src/Sylius/Bundle/ApiBundle/test/config/api_resources/Promotion.xml [hunk 20]
```diff
@@ -0,0 +1,58 @@
+x<?xml version="1.0" ?>
```
Comment (stloyd) [inline src/Sylius/Bundle/ApiBundle/test/config/api_resources/Promotion.xml:1]:
```suggestion
<?xml version="1.0" ?>
```

**Review by arti0090 — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/test/config/fixtures/foo.yaml [hunk 21]
```diff
@@ -0,0 +1,13 @@
+Sylius\Bundle\ApiBundle\Application\Entity\FooSyliusResource:
+    fooSyliusResource3:
+        name: FooSyliusResource3
+
+Sylius\Bundle\ApiBundle\Application\Entity\Foo:
+    foo1:
+        name: Foo0
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/test/config/fixtures/foo.yaml:7]: can we synchronize this naming?

**Review by lchrusciel — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/test/config/fixtures/foo.yaml (hunk 21 above)
Comment (GSadee) [reply src/Sylius/Bundle/ApiBundle/test/config/fixtures/foo.yaml:7]: Definitely, we should, but I would like to do that in a separate PR as it requires a little refactor of unit tests

**Comment by lchrusciel** [comment]
Thank you, Tomasz! :1st_place_medal:

**Pull request merged**